### DISCUSSION TITLE
Elsa in a Windows service?

### DISCUSSION THREAD

**Discussion by mrubelmann** [elsa-workflows/elsa-core#2658] [Q&A] [answered]
Is it possible to run Elsa workflows in a Windows service?  My preference would be to implement the service with the Microsoft.Extensions.Hosting.WindowsServices package and then manually run the workflows, but it seems like Elsa isn't really set up with that in mind.  Can that be done easily or is it an uphill battle?

My other assumption was that hosting it in a Windows service wouldn't be much different than ASP.NET, but I see all the samples are using `Host.CreateDefaultBuilder(args).ConfigureWebHostDefaults(webBuilder => { webBuilder.UseStartup<Startup>(); })`, and I can't find an equivalent construct in Microsoft.Extensions.Hosting.WindowsServices.  Is there something I'm missing?

**Comment by sfmskywalker** [comment]
Yes, Elsa can run in any host, including in a windows service. But you’re right, we don’t have any samples of that yet. Simply because I personally never had the need for it, but it’s definitely a legit use case.

**Reply by mrubelmann** [reply]
Thanks Sipke!  When you say it can run in any host, do you mean I should be able to start it up by calling `AddElsa()` or can I use the service I have now and manually start workflows?  I've tried letting Elsa run the show (with `AddElsa()`) instead of using our current service class (which inherits from `BackgroundService`), but so far I haven't been able to get Windows to realize the service is actually running.

**Comment by mrubelmann** [comment, marked as answer]
I figured out a solution - run Elsa in a scoped service within the primary service.  You can set up both your primary service and Elsa at the start of the program, making sure to include `.UseWindowsService()` to make it runnable as a full fledged Windows service.

```static async Task Main(string[] args)
{
    Host.CreateDefaultBuilder(args)
        .ConfigureLogging(configureLogging => configureLogging.AddNLog())
        .ConfigureServices((hostContext, services) => services
            .AddHostedService<CsService>()
            .AddElsa(options => options
                .AddConsoleActivities()
                .AddWorkflow<TestWorkflow>()
            )
        )
        .UseWindowsService()
        .Build()
        .Run();
}
```

.NET runs the primary service, which then starts the Elsa workflow:

```public class CsService : BackgroundService
{
    private readonly IServiceProvider _serviceProvider;
    private readonly ILogger<CsService> Log;

    public CsService(IServiceProvider serviceProvider, ILogger<CsService> logger)
    {
        _serviceProvider = serviceProvider;
        Log = logger;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        Log.LogInformation("Service Starting");

        var tcs = new TaskCompletionSource<bool>();
        stoppingToken.Register(s => ((TaskCompletionSource<bool>?)s)?.SetResult(true), tcs);

        using (var scope = _serviceProvider.CreateScope())
        {
            var workflowRunner = scope.ServiceProvider.GetRequiredService<IBuildsAndStartsWorkflow>();
            await workflowRunner.BuildAndStartWorkflowAsync<TestWorkflow>();
        }

        await tcs.Task;
        
        Log.LogInformation("Service Stopping");
    }
}
```

That lets Windows realize that the service came up successfully and you can start/stop it as usual.
